modules/auxiliary/scanner/msmail/user_enum.go

```diff
@@ -0,0 +1,346 @@
+		medianTime = (sliceOfTimes[positionTwo] + sliceOfTimes[positionOne]) / 2
+	} else if len(sliceOfTimes)%2 != 0 {
+		position := len(sliceOfTimes)/2 - 1
+		medianTime = sliceOfTimes[position]
+	} else {
+		module.LogError("Error determining whether length of times gathered is even or odd to obtain median value.")
+	}
+	module.LogInfo("Avg Response: " + string(time.Duration(medianTime)))
+	return time.Duration(medianTime)
+}
+
+func o365enum(emaillist []string, threads int) []string {
+	limit := threads
+	var wg sync.WaitGroup
+	mux := &sync.Mutex{}
+	queue := make(chan string)
+	//limit := 100
+
+	/*Keep in mind you, nothing has been added to handle successful auths
+	  so the password for auth attempts has been hardcoded to something
+	  that is not likely to be correct.
+	*/
+	pass := "Summer2018876"
+	URI := "https://outlook.office365.com/Microsoft-Server-ActiveSync"
+	var validemails []string
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+
+	for i := 0; i < limit; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			for email := range queue {
+				responseCode := webRequestBasicAuth(URI, email, pass, tr)
+				if strings.Contains(email, "@") && responseCode == 401 {
+					mux.Lock()
+					module.LogInfo("[+]  " + email + " - 401")
+					validemails = append(validemails, email)
+					mux.Unlock()
+				} else if strings.Contains(email, "@") && responseCode == 404 {
+					mux.Lock()
+					module.LogInfo(fmt.Sprintf("[-]  %s - %d \n", email, responseCode))
+					mux.Unlock()
+				} else {
+					mux.Lock()
+					module.LogInfo(fmt.Sprintf("Unusual Response: %s - %d \n", email, responseCode))
+					mux.Unlock()
+				}
+			}
+		}(i)
+	}
+
+	for i := 0; i < len(emaillist); i++ {
+		queue <- emaillist[i]
+	}
+
+	close(queue)
+	wg.Wait()
+	return validemails
+}
+
+func webRequestBasicAuth(URI string, user string, pass string, tr *http.Transport) int {
+	timeout := time.Duration(45 * time.Second)
+	client := &http.Client{
+		Timeout:   timeout,
+		Transport: tr,
+	}
+	req, err := http.NewRequest("GET", URI, nil)
+	req.Header.Set("User-Agent", "Mozilla/5.0 (iPhone; CPU iPhone OS 11_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1")
+	req.SetBasicAuth(user, pass)
+	resp, err := client.Do(req)
+	if err != nil {
+		module.LogInfo(fmt.Sprintf("Potential Timeout - %s \n", user))
+		module.LogInfo("One of your requests has taken longer than 45 seconds to respond.")
+		module.LogInfo("Consider lowering amount of threads used for enumeration.")
+		module.LogError(err.Error())
+	}
+	return resp.StatusCode
+}
+
+func reportValidUsers(ip string, validUsers []string) {
+	port := "443"
+	service := "owa"
+	protocol := "tcp"
+	for _, user := range validUsers {
+		opts := map[string]string{
+			"port":         port,
+			"service_name": service,
+			"address":         ip,
+			"protocol":		protocol,
+			"fullname": 	"auxiliary/scanner/msmail/msmail_enum",
```

Comment:
Is there a way to determine the fullname programmatically?

Reply:
if you figure it out let me know :)

Reply:
The report method should not need this. The `Msf::Module::External` mixin has access to this info and should be the one using it